### ISSUE TITLE
Add nock tests for handlers

### ISSUE THREAD

**Issue by damccorm** [microsoft/typed-rest-client#90]
Should test all of them, but especially NTLM since it doesn't do preauth and is pretty fragile

**Comment by hamzahejja** [comment]
@damccorm I'm able to pre-plan the unit tests for the Basic, Personal Access Token and Bearer Token Authentication schemes, as they all perform pre-auth by Setting the `Authorization` header on the outgoing request. However, couldn't imagine how to carry-on the test for the NTLM auth, especially as it involves CLIENT-SERVER Handshaking and is quite complex compared to other schemes, So i bumped into a wall while thinking to write the test. Any suggestions? thoughts ?

**Comment by hamzahejja** [comment]
@damccorm I've managed to pre-plan and setup the needed tests for **_Basic, Bearer and Personal Access Token (PAT)_** Authentication schemes, by setting nock to match the `Authorization` header on the outgoing HTTP request. However, I'm wondering how to carry out the test for the NTLM Authentication scheme, as it is more complex, doesn't perform any Pre-auth step on the request and involves Client-Server Handshaking. Any thoughts on this? suggestions?

**Comment by damccorm** [comment]
@hamzahejja the NTLM one is definitely the hardest - I think you should still be able to do it more or less the same way though, right? You would just need to mock multiple sets of HTTP requests with different parameters, which nock _should_ allow.

With that said, feel free to open a PR on the other 3 handlers if you have them - happy to get those merged in and we can try to work on the ntlm one separately.

**Comment by hamzahejja** [comment]
@damccorm I have just had another look into the tests which currently exists,  **PAT and Basic** Authentication schemes have nock tests set for them:
_Basic Auth_: https://github.com/microsoft/typed-rest-client/blob/master/test/units/httptests.ts#L72-L99
_Personal Access Token (PAT)_: https://github.com/microsoft/typed-rest-client/blob/master/test/units/httptests.ts#L130-L160

Which leaves out the Bearer Token Auth. scheme and the NTLM, right ?

**Comment by damccorm** [comment]
Yep, that sounds right!

**Comment by hamzahejja** [comment]
@damccorm
I have set the following nock test for the NTLM Authentication Handler, depending on the example given **_BOTTOM OF PAGE_**, with the following settings:
`username`: **Zaphod**
`password`: **Beeblebrox**
`workstation` (i.e. hostname): **LightCity**
`domain` (i.e. NTDomain): **Ursa-Minor**
```
        const url: string = 'http://microsoft.com';
        const base64EncodedType1Message = 'NTLM TlRMTVNTUAABAAAAA7IAAAoACgApAAAACQAJACAAAABMSUdIVENJVFlVUlNBLU1JTk9S';
        const base64EncodedType2Message = 'NTLM TlRMTVNTUAACAAAAAAAAACgAAAABggAAU3J2Tm9uY2UAAAAAAAAAAA==';
        const base64EncodedType3Message = 'NTLM TlRMTVNTUAADAAAAGAAYAHIAAAAYABgAigAAABQAFABAAAAADAAMAFQAAAASABIAYAAAAAAAAACiAAAAAYIAAFUAUgBTAEEALQBNAEkATgBPAFIAWgBhAHAAaABvAGQATABJAEcASABUAEMASQBUAFkArYfKbe/jRoW5xDxHeoxC1gBmfWiS5+iX4OAN4xBKG/IFPwfH3agtPEia6YnhsADT';

        const ntlmScopeType1 = nock(url)
            .get('/')
            .reply(httpm.HttpCodes.Unauthorized, {}, {'WWW-Authenticate': 'NTLM'});

        const ntlmScopeType2 = nock(url)
            .matchHeader('Connection', 'keep-alive')
            .matchHeader('Authorization', base64EncodedType1Message)
            .get('/')
            .reply(httpm.HttpCodes.Unauthorized, {}, {'WWW-Authenticate': base64EncodedType2Message});

        const ntlmScopeType3 = nock(url)
            .matchHeader('Connection', 'Close')
            .matchHeader('Authorization', base64EncodedType3Message)
            .get('/')
            .reply(httpm.HttpCodes.OK, {success: true});

        const ntlmAuthHandler: hm.NtlmCredentialHandler = new hm.NtlmCredentialHandler('Zaphod', 'Beeblebrox', 'LightCity', 'Ursa-Minor');
        const httpClient: httpm.HttpClient = new httpm.HttpClient(undefined, [ntlmAuthHandler]);
        const httpResponse: httpm.HttpClientResponse = await httpClient.get(url);
        const jsonBody = await httpResponse.readBody().then(JSON.parse)

        assert(ntlmScopeType1.isDone() && ntlmScopeType2.isDone() && ntlmScopeType3.isDone());
        assert(httpResponse.message.statusCode == httpm.HttpCodes.OK);
        assert(jsonBody.success);
```
The issue arises when using the NTLM Authentication Handler with HTTP Client, and was noticed during setting up a nock test for the NTLM Handler. The problem is the NTLM messages produced, since it seems the current codebase which was built with the help of: [SamDecrock's http-ntlm lib] (https://github.com/SamDecrock/node-http-ntlm/blob/master/ntlm.js) produces incorrect values for type1 message, which will eventually lead to producing an incorrectly-encoded type3 message at the end of the NTLM handshake.

### Expected behaviour
- Per [NTLM Authentication Scheme for HTTP](https://www.innovation.ch/personal/ronald/ntlm.html), Observing the Example at **bottom of page**, with `workstation` set to _LightCity_ and `NTdomain` set to `Ursa-Minor`, base64-encoded Type1 message should be: `NTLM TlRMTVNTUAABAAAAA7IAAAoACgApAAAACQAJACAAAABMSUdIVENJVFlVUlNBLU1JTk9S`.

![image](https://user-images.githubusercontent.com/49674728/71614938-32d43900-2bb7-11ea-9b14-775542e7f8de.png)


### Actual behaviour
- Using `workstation` = **LightCity** and `domain` = **Ursa-Minor** based on the above Example from [NTLM Authentication Scheme for HTTP](https://www.innovation.ch/personal/ronald/ntlm.html), our current codebase for the NTLM handler produces `NTLM TlRMTVNTUAABAAAAB7IIogoACgAxAAAACQAJACgAAAAFASgKAAAAD0xJR0hUQ0lUWVVSU0EtTUlOT1I=` for the type1 message, as the request expected this type1 message! 

```
Error: Nock: No match for request {
  "method": "GET",
  "url": "http://microsoft.com/",
  "headers": {
    "connection": "keep-alive",
    "authorization": "NTLM TlRMTVNTUAABAAAAB7IIogoACgAxAAAACQAJACgAAAAFASgKAAAAD0xJR0hUQ0lUWVVSU0EtTUlOT1I="
  }
```

Could this be an issue with the functions used for NTLM Auth Handler ?

**Comment by hamzahejja** [comment]
@damccorm I've found this open source lib for NTLM Authentication with NodeJS: [node-ntlm, NTLM authentication and Samba LM/NT hash library](https://github.com/Node-SMB/ntlm)

Tests for the module cover the [example](https://www.innovation.ch/personal/ronald/ntlm.html), and returns matching values indeed! 
![image](https://user-images.githubusercontent.com/49674728/71632008-bbcd8d80-2c14-11ea-9cf9-85ef6f3e5e24.png)


How about changing the NTLM's handler code base to using the above mentioned module's functions? With writing the required nock test for the NTLM handler, sounds good ?

**Comment by damccorm** [comment]
> How about changing the NTLM's handler code base to using the above mentioned module's functions? Afterwards, I'll write the required nock test for the NTLM handler to ensure working functionality, sounds legit ?

That sounds good to me. My one request here - for auditability/security, I'd rather not pull their module directly from npm since its not a trusted author - could you copy in only the required functions and attribute them to the author (which should be fine as long as we include [this attribution](https://github.com/Node-SMB/ntlm/blob/126971e45ad4a84025c13291eddfb32f1288c5e5/lib/ntlm.js#L2)). That way all the new code gets reviewed and we have a little more control over things like security patches

**Comment by hamzahejja** [comment]
@damccorm Yeah totally agree on that as well, just what I've thought too, to just copy in the files that we'll be using through the NTLM authentication handler module and include an attribution to the author's work. I'm really curious why [Sam Decrock's NTLM lib](https://github.com/SamDecrock/node-http-ntlm/blob/master/ntlm.js) isn't making it work, although in means of Github Stars and People using the library, forks, etc .. it seems highly used and should do the job! What makes things weird more, is the fact that the lib mentions the [Example](https://www.innovation.ch/personal/ronald/ntlm.html) in their [GitHub page](https://github.com/SamDecrock/node-http-ntlm#more-information). So we got green light on the new module to cover the NTLM auth. handler ?

**Comment by hamzahejja** [comment]
Found another example to follow-on and ensure correct authentication headers/codes are produced for NTLM Authentication handler while establishing the C-S handshake, find it here: [NTLM HTTP Authentication
](http://davenport.sourceforge.net/ntlm.html#ntlmHttpAuthentication)

**Comment by AnthonyDiSanti** [comment]
I think this issue can be closed now that @damccorm has merged #194 .  Does that sound right to you as well @hamzahejja ?

**Comment by damccorm** [comment]
I think that's right - I'm going to close, if anyone disagrees comment and we can reopen.
